internal/handlers/pg/msg_aggregate.go

```diff
@@ -243,40 +258,28 @@ func (h *Handler) MsgAggregate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMs
 
 	var iter iterator.Interface[struct{}, *types.Document]
 
-	// At this point we have a list of stages to apply to the documents or stats.
-	// If collStatsDocuments contains the same stages as stagesDocuments, we apply aggregation to documents fetched from the DB.
-	// If collStatsDocuments contains more stages than stagesDocuments, we apply aggregation to statistics fetched from the DB.
-	if len(collStatsDocuments) == len(stagesDocuments) {
-		filter, sort := aggregations.GetPushdownQuery(aggregationStages)
-
-		// only documents stages or no stages - fetch documents from the DB and apply stages to them
-		qp := pgdb.QueryParams{
-			DB:         db,
-			Collection: collection,
-		}
+	if len(producerStages) == 0 {
+		var s aggregations.ProducerStage
 
-		if !h.DisableFilterPushdown {
-			qp.Filter = filter
+		s, err = stages.NewDocumentProducer(p)
+		if err != nil {
+			return nil, err
 		}
 
-		if h.EnableSortPushdown {
-			qp.Sort = sort
+		if iter, err = s.Produce(ctx, closer); err != nil {
+			return nil, lazyerrors.Error(err)
 		}
-
-		iter, err = processStagesDocuments(ctx, closer, &stagesDocumentsParams{dbPool, &qp, stagesDocuments})
 	} else {
-		// stats stages are provided - fetch stats from the DB and apply stages to them
-		// TODO move $collStatsDocuments specific logic to its stage https://github.com/FerretDB/FerretDB/issues/2423
-		statistics := stages.GetStatistics(collStatsDocuments)
-
-		iter, err = processStagesStats(ctx, closer, &stagesStatsParams{
-			dbPool, db, collection, statistics, collStatsDocuments,
-		})
+		// handle case where there are multiple producer stages
```

Comment:
Is it a possible situation - multiple producer stages?

Reply:
No

Reply:
For example, having two consecutive `$collStats` stage should return error.

Reply:
That's correct, but that's not what this code does, isn't it?

I think we should return an error and exit early waaay up when we iterate over `aggregationStages`